src/Microsoft.Health.Fhir.TemplateManagement/ArtifactProviders/BlobTemplateProvider.cs

```diff
@@ -0,0 +1,90 @@
+﻿// -------------------------------------------------------------------------------------------------
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License (MIT). See LICENSE in the repo root for license information.
+// -------------------------------------------------------------------------------------------------
+
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+using Azure;
+using Azure.Storage.Blobs;
+using Azure.Storage.Blobs.Models;
+using DotLiquid;
+using Microsoft.Extensions.Caching.Memory;
+using Microsoft.Health.Fhir.Liquid.Converter.Utilities;
+using Microsoft.Health.Fhir.TemplateManagement.Models;
+
+namespace Microsoft.Health.Fhir.TemplateManagement.ArtifactProviders
+{
+    public class BlobTemplateProvider : IConvertDataTemplateProvider
+    {
+        private BlobContainerClient _blobContainerClient;
+
+        private readonly IMemoryCache _templateCache;
+
+        private readonly TemplateCollectionConfiguration _templateCollectionConfiguration;
+
+        private readonly string _blobTemplateCacheKey = "cached-blob-templates";
+
+        private readonly int _segmentSize = 100;
+
+        public BlobTemplateProvider(BlobContainerClient blobContainerClient, IMemoryCache templateCache, TemplateCollectionConfiguration templateConfiguration)
+        {
+            _blobContainerClient = blobContainerClient;
+            _templateCache = templateCache;
+            _templateCollectionConfiguration = templateConfiguration;
+        }
+
+        public async Task<List<Dictionary<string, Template>>> GetTemplateCollectionAsync(CancellationToken cancellationToken = default)
+        {
+            // read templates from cache if available
+            if (_templateCache.TryGetValue(_blobTemplateCacheKey, out var templateCache))
+            {
+                return (List<Dictionary<string, Template>>)templateCache;
```

Comment:
Not sure about a the public contract returning the entire cache, especially since the we would want to use the IMemoryCache to enforce the lifecycle of the cached item.  Once the dictionary is returned we don't control what is done with the object and it can potentially out live the cache policy we set.  I think a better strategy would be hide that detail from the caller and only return the exact template needed and like the internal logic determine the result.

Another question, what is the cache key in this instance?  Is it the storage location or the individual template?  I am trying to reconcile why inside the cache is a dictionary of templates rather than a single template.

All that said, I realize some of this is due to the existing implementation.  I am fine deferring these changes until we look a unifying the template logic across sources.